File: src/services/MarkdownFormatter.ts
import type {
  DomainExploration,
  FormalityAlternative,
  MistakeItem,
  PracticeExercise,
  WordSection1Header,
  WordSection2Meanings,
  WordSection3Related,
  UsagePattern,
} from '../types';
import { ERROR_MESSAGES } from '../utils/errorHandler';

export interface WordExportPayload {
  header?: WordSection1Header | null;
  meanings?: WordSection2Meanings | null;
  related?: WordSection3Related | null;
  exploration?: ExplorationExportPayload | null;
}

export interface ExplorationExportPayload {
  formality?: {
    percentage?: number | null;
    alternatives?: FormalityAlternative[];
  };
  domains?: DomainExploration[];
  usage?: UsagePattern[];
  practice?: PracticeExercise[];
  mistakes?: MistakeItem[];
  customContext?: {
    label?: string;
    examples: string[];
  };
}

export interface MarkdownFormatOptions {
  includeTimestamp?: boolean;
  tags?: string[];
  includeExploration?: boolean;
}

const ensureHeader = (payload: WordExportPayload): WordSection1Header => {
  if (!payload.header) {
    throw new Error(ERROR_MESSAGES.EXPORT_NOT_READY);
  }
  return payload.header;
};

/**
 * Formats word data into markdown with YAML frontmatter.
 * Single Responsibility: Markdown document generation only.
 */
export class MarkdownFormatter {
  static format(
    payload: WordExportPayload,
    options: MarkdownFormatOptions = {}
  ): string {
    const header = ensureHeader(payload);
    const lines: string[] = [];

    // Frontmatter metadata
    lines.push('---');
    lines.push(`word: ${header.word}`);
    lines.push(`date: ${new Date().toISOString().split('T')[0]}`);
    lines.push(`formality: ${header.formality.percentage}%`);
    lines.push(`origin: ${header.origin}`);
    if (options.tags?.length) {
      const tags = options.tags.map(t => t.trim().replace(/^#/, '')).filter(Boolean);
      lines.push(`tags: [${tags.join(', ')}]`);
    }
    lines.push('---');
    lines.push('');

    // Title and metadata
    lines.push(`# ${header.word}`);
    lines.push('');
    lines.push(`**Pronunciation:** /${header.pronunciation}/`);
    lines.push(`**Part of Speech:** ${payload.meanings?.meanings?.[0]?.partOfSpeech || 'N/A'}`);
    lines.push(`**Formality:** ${header.formality.percentage}% ${header.formality.level}`);
    lines.push(`**Origin:** ${header.origin}`);
    lines.push('');
    lines.push('---');
    lines.push('');

    // TL;DR
    lines.push('## TL;DR');
    lines.push('');
    lines.push(`> **${header.tldr}**`);
    lines.push('');
    lines.push('---');
    lines.push('');

    // Meanings
    if (payload.meanings?.meanings?.length) {
      lines.push('## Meanings');
      lines.push('');
      payload.meanings.meanings.forEach((meaning) => {
        lines.push(`### ${meaning.number}. ${meaning.definition}`);
        lines.push('');
        if (meaning.examples?.length) {
          meaning.examples.forEach((example) => {
            lines.push(`${example}`);
          });
          lines.push('');
        }
        if (meaning.memoryTip) {
          lines.push(`**Memory Tip:** ${meaning.memoryTip}`);
          lines.push('');
        }
      });
      lines.push('---');
      lines.push('');
    }

    // Vocabulary Network
    const hasRelated = payload.related?.synonyms?.length || 
                       payload.related?.antonyms?.length || 
                       payload.related?.collocations?.length;
    
    if (hasRelated) {
      lines.push('## Vocabulary Network');
      lines.push('');
      
      if (payload.related?.synonyms?.length) {
        lines.push(`**Synonyms**`);
        lines.push(payload.related.synonyms.join(' • '));
        lines.push('');
      }
      
      if (payload.related?.antonyms?.length) {
        lines.push(`**Antonyms**`);
        lines.push(payload.related.antonyms.join(' • '));
        lines.push('');
      }
      
      if (payload.related?.collocations?.length) {
        lines.push(`**Collocations**`);
        payload.related.collocations.forEach((collocation) => {
          lines.push(`- ${collocation.phrase}`);
        });
        lines.push('');
      }
      
      lines.push('---');
      lines.push('');
    }

    // Formality Guide
    if (options.includeExploration && payload.exploration?.formality) {
      const { alternatives } = payload.exploration.formality;
      if (alternatives && alternatives.length > 0) {
        lines.push('## Formality Guide');
        lines.push('');
        lines.push('| Formality | Alternative | Context |');
        lines.push('|-----------|-------------|---------|');
        
        alternatives.forEach((alt) => {
          const register = alt.level || '';
          const alternative = alt.word || '';
          const context = alt.context || '';
          lines.push(`| **${register}** | ${alternative} | ${context} |`);
        });
        
        lines.push('');
        lines.push('---');
        lines.push('');
      }
    }

    // Domain-Specific Usage
    if (options.includeExploration && payload.exploration?.domains?.length) {
      lines.push('## Domain-Specific Usage');
      lines.push('');
      
      payload.exploration.domains.forEach((domain) => {
        lines.push(`### ${domain.domain}`);
        if (domain.examples?.length) {
          domain.examples.forEach((example) => {
            lines.push(`"${example}"`);
          });
        }
        lines.push('');
      });
      
      lines.push('---');
      lines.push('');
    }

    // Common Pitfalls
    if (options.includeExploration && payload.exploration?.mistakes?.length) {
      lines.push('## Common Pitfalls');
      lines.push('');
      
      payload.exploration.mistakes.forEach((mistake) => {
        lines.push(`**Incorrect (${mistake.type}):** ${mistake.incorrectUsage}`);
        lines.push(`**Correct:** ${mistake.correction}`);
        lines.push('');
      });
      
      lines.push('---');
      lines.push('');
    }

    // Usage Patterns
    if (options.includeExploration && payload.exploration?.usage?.length) {
      lines.push('## Grammar Patterns');
      lines.push('');
      
      payload.exploration.usage.forEach((pattern) => {
        lines.push(`**${pattern.template}**`);
        lines.push(`${pattern.description}`);
        if (pattern.examples?.length) {
          pattern.examples.forEach((example) => {
            lines.push(`- ${example}`);
          });
        }
        lines.push('');
      });
      
      lines.push('---');
      lines.push('');
    }

    // Practice
    if (options.includeExploration && payload.exploration?.practice?.length) {
      lines.push('## Practice');
      lines.push('');
      
      payload.exploration.practice.forEach((exercise, index) => {
        lines.push(`**${index + 1}. ${exercise.question}**`);
        lines.push('');
        exercise.options.forEach((option) => {
          const marker = option === exercise.correctAnswer ? 'x' : ' ';
          lines.push(`   [${marker}] ${option}`);
        });
        if (exercise.explanation) {
          lines.push('');
          lines.push(`   *${exercise.explanation}*`);
        }
        lines.push('');
      });
      
      lines.push('---');
      lines.push('');
    }

    // Footer
    lines.push('---');
    lines.push('');
    const timestamp = options.includeTimestamp !== false ? 
      new Date().toISOString().split('T')[0] : '';
    lines.push(`*Generated by MelliLex${timestamp ? ' · ' + timestamp : ''}*`);

    return lines.join('\n').trim() + '\n';
  }
}

File: src-tauri/src/export/markdown_formatter.rs
use crate::models::{
    PhraseDefinitionData, PhraseSection1Overview, PhraseSection2Context, PhraseSection3Related,
    WordProgressiveData, WordSection1Header, WordSection2Meanings, WordSection3Related,
};

/// Single Responsibility: Format WordProgressiveData into markdown text
pub struct MarkdownFormatter;

impl MarkdownFormatter {
    pub fn format(word: &str, data: &WordProgressiveData, include_timestamp: bool) -> String {
        let mut md = String::new();

        // Title
        md.push_str(&format!("# {}\n\n", word));

        // Section 1: Header
        md.push_str(&Self::format_header_section(&data.section1));
        md.push_str("\n\n");

        // Section 2: Meanings
        md.push_str(&Self::format_meanings_section(&data.section2));
        md.push_str("\n\n");

        // Section 3: Related Words
        md.push_str(&Self::format_related_section(&data.section3));

        // Optional timestamp
        if include_timestamp {
            md.push_str("\n\n---\n");
            md.push_str(&format!(
                "*Exported: {}*\n",
                chrono::Utc::now().format("%Y-%m-%d %H:%M:%S UTC")
            ));
        }

        md
    }

    fn format_header_section(header: &WordSection1Header) -> String {
        let mut md = String::new();

        // Pronunciation and basic info
        if !header.pronunciation.is_empty() {
            md.push_str(&format!("**Pronunciation:** {}\n\n", header.pronunciation));
        }

        if !header.syllables.is_empty() {
            md.push_str(&format!("**Syllables:** {}\n\n", header.syllables));
        }

        // TL;DR
        if !header.tldr.is_empty() {
            md.push_str("## In Simple Words\n\n");
            md.push_str(&format!("{}\n\n", header.tldr));
        }

        // Formality
        md.push_str(&format!(
            "**Formality:** {} ({}%)\n\n",
            header.formality.level, header.formality.percentage
        ));

        // Domains
        if !header.domains.is_empty() {
            md.push_str(&format!("**Used in:** {}\n\n", header.domains.join(", ")));
        }

        // Origin
        if !header.origin.is_empty() {
            md.push_str(&format!("**Origin:** {}\n", header.origin));
        }

        md
    }

    fn format_meanings_section(meanings: &WordSection2Meanings) -> String {
        let mut md = String::new();
        md.push_str("## Meanings\n\n");

        for meaning in &meanings.meanings {
            md.push_str(&format!(
                "### {}. {} ({})\n\n",
                meaning.number, meaning.definition, meaning.part_of_speech
            ));

            // Memory tip
            if !meaning.memory_tip.is_empty() {
                md.push_str(&format!("**Memory Tip:** {}\n\n", meaning.memory_tip));
            }

            // Examples
            if !meaning.examples.is_empty() {
                md.push_str("**Examples:**\n\n");
                for example in &meaning.examples {
                    md.push_str(&format!("- {}\n", example));
                }
                md.push('\n');
            }
        }

        md
    }

    fn format_related_section(related: &WordSection3Related) -> String {
        let mut md = String::new();
        md.push_str("## Related Words\n\n");

        // Synonyms
        if !related.synonyms.is_empty() {
            md.push_str("**Synonyms:** ");
            md.push_str(&related.synonyms.join(", "));
            md.push_str("\n\n");
        }

        // Antonyms
        if !related.antonyms.is_empty() {
            md.push_str("**Antonyms:** ");
            md.push_str(&related.antonyms.join(", "));
            md.push_str("\n\n");
        }

        // Collocations
        if !related.collocations.is_empty() {
            md.push_str("**Common Collocations:**\n\n");
            for collocation in &related.collocations {
                md.push_str(&format!(
                    "- **{}**: {}\n",
                    collocation.phrase, collocation.example
                ));
            }
        }

        md
    }

    // ============== PHRASE FORMATTING ==============

    pub fn format_phrase(
        phrase: &str,
        data: &PhraseDefinitionData,
        include_timestamp: bool,
    ) -> String {
        let mut md = String::new();

        // Title
        md.push_str(&format!("# {}\n\n", phrase));

        // Section 1: Overview
        md.push_str(&Self::format_phrase_overview(&data.section1));
        md.push('\n');

        // Section 2: Context
        md.push_str(&Self::format_phrase_context(&data.section2));
        md.push('\n');

        // Section 3: Related
        md.push_str(&Self::format_phrase_related(&data.section3));

        // Optional timestamp
        if include_timestamp {
            md.push_str("\n---\n");
            md.push_str(&format!(
                "*Exported: {}*\n",
                chrono::Utc::now().format("%Y-%m-%d %H:%M:%S UTC")
            ));
        }

        md
    }

    fn format_phrase_overview(overview: &PhraseSection1Overview) -> String {
        let mut md = String::new();

        // Metadata line
        md.push_str(&format!(
            "**Type:** {} · **Region:** {} · **Formality:** {}\n\n",
            overview.phrase_type, overview.region, overview.formality.level
        ));

        // TL;DR
        md.push_str("## In a Nutshell\n\n");
        md.push_str(&format!("{}\n\n", overview.tldr));

        // Literal meaning (if exists)
        if let Some(ref literal) = overview.literal_meaning {
            md.push_str("## Literal Meaning\n\n");
            md.push_str(&format!("{}\n\n", literal));
        }

        // Actual meaning
        md.push_str("## Actual Meaning\n\n");
        md.push_str(&format!("{}\n", overview.actual_meaning));

        md
    }

    fn format_phrase_context(context: &PhraseSection2Context) -> String {
        let mut md = String::new();

        // Origin Story
        md.push_str("## Origin Story\n\n");
        if let Some(ref era) = context.origin.era {
            md.push_str(&format!("**Era:** {}\n\n", era));
        }
        if let Some(ref source) = context.origin.source {
            md.push_str(&format!("**Source:** {}\n\n", source));
        }
        md.push_str(&format!("{}\n", context.origin.story));
        if let Some(ref evolution) = context.origin.evolution {
            md.push_str(&format!("\n**Evolution:** {}\n", evolution));
        }
        md.push('\n');

        // Usage Notes
        if !context.usage_notes.is_empty() {
            md.push_str("## Usage Notes\n\n");
            for note in &context.usage_notes {
                let tone_str = note
                    .tone
                    .as_ref()
                    .map(|t| format!(" ({})", t))
                    .unwrap_or_default();
                md.push_str(&format!("### {}{}\n\n", note.context, tone_str));
                md.push_str(&format!("> {}\n\n", note.example));
            }
        }

        // Common Mistakes
        if !context.common_mistakes.is_empty() {
            md.push_str("## Common Mistakes\n\n");
            for mistake in &context.common_mistakes {
                md.push_str(&format!("### {}\n\n", mistake.mistake_type));
                md.push_str(&format!("Incorrect: {}\n\n", mistake.incorrect));
                md.push_str(&format!("Correct: {}\n\n", mistake.correct));
                md.push_str(&format!("> {}\n\n", mistake.explanation));
            }
        }

        md
    }

    fn format_phrase_related(related: &PhraseSection3Related) -> String {
        let mut md = String::new();

        // Variations
        if !related.variations.is_empty() {
            md.push_str("## Variations\n\n");
            for variation in &related.variations {
                let region_str = variation
                    .region
                    .as_ref()
                    .map(|r| format!(" ({})", r))
                    .unwrap_or_default();
                let note_str = variation
                    .note
                    .as_ref()
                    .map(|n| format!(" — {}", n))
                    .unwrap_or_default();
                md.push_str(&format!(
                    "- **{}**{}{}\n",
                    variation.phrase, note_str, region_str
                ));
            }
            md.push('\n');
        }

        // Similar Phrases
        if !related.similar_phrases.is_empty() {
            md.push_str("## Similar Phrases\n\n");
            for phrase in &related.similar_phrases {
                md.push_str(&format!("- **{}** — {}\n", phrase.phrase, phrase.meaning_hint));
            }
            md.push('\n');
        }

        // Opposite Phrases
        if !related.opposite_phrases.is_empty() {
            md.push_str("## Opposite Phrases\n\n");
            for phrase in &related.opposite_phrases {
                md.push_str(&format!("- **{}** — {}\n", phrase.phrase, phrase.meaning_hint));
            }
            md.push('\n');
        }

        // See Also
        if !related.see_also.is_empty() {
            md.push_str("## See Also\n\n");
            for item in &related.see_also {
                md.push_str(&format!("- {}\n", item));
            }
            md.push('\n');
        }

        md
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::*;

    fn create_test_data() -> WordProgressiveData {
        WordProgressiveData {
            section1: WordSection1Header {
                word: "ephemeral".to_string(),
                pronunciation: "/ɪˈfɛm.ə.rəl/".to_string(),
                syllables: "e·phem·er·al".to_string(),
                origin: "Greek ephēmeros 'lasting only a day'".to_string(),
                formality: FormalityInfo {
                    level: "Formal".to_string(),
                    percentage: 75,
                },
                domains: vec!["Literature".to_string(), "Philosophy".to_string()],
                tldr: "Something that lasts for a very short time".to_string(),
            },
            section2: WordSection2Meanings {
                meanings: vec![MeaningItem {
                    number: 1,
                    part_of_speech: "adjective".to_string(),
                    definition: "Lasting for a very short time".to_string(),
                    memory_tip: "Think of 'ephemeral' as something that disappears quickly, like morning dew".to_string(),
                    examples: vec![
                        "The beauty of cherry blossoms is ephemeral.".to_string(),
                    ],
                }],
            },
            mistakes: None,
            section3: WordSection3Related {
                synonyms: vec!["transient".to_string(), "fleeting".to_string()],
                antonyms: vec!["permanent".to_string(), "enduring".to_string()],
                collocations: vec![CollocationItem {
                    phrase: "ephemeral nature".to_string(),
                    example: "The ephemeral nature of fame".to_string(),
                }],
            },
        }
    }

    #[test]
    fn test_format_markdown() {
        let data = create_test_data();
        let md = MarkdownFormatter::format("ephemeral", &data, false);

        assert!(md.contains("# ephemeral"));
        assert!(md.contains("**Pronunciation:**"));
        assert!(md.contains("## Meanings"));
        assert!(md.contains("## Related Words"));
    }
}
